craft_providers/bases/instance_config.py

```diff
@@ -18,94 +18,135 @@
-            )
-        except subprocess.CalledProcessError:
-            return None
-
-        try:
-            proc = executor.execute_run(
-                command=["cat", config_path.as_posix()],
-                capture_output=True,
-                check=True,
-                text=True,
-            )
-        except subprocess.CalledProcessError as error:
+            return InstanceConfiguration(**data)
+        except pydantic.ValidationError as error:
             raise BaseConfigurationError(
-                brief=f"Failed to read instance config in environment at {config_path.as_posix()}",
-                details=errors.details_from_called_process_error(error),
+                brief="Failed to parse instance configuration file.",
+                details=f"Instance configuration data: {data!r}",
             ) from error
 
-        data = yaml.safe_load(proc.stdout)
+    def marshal(self) -> Dict[str, Any]:
+        """Create a dictionary containing the InstanceConfiguration data.
 
-        try:
-            return cls.unmarshal(data)
-        except (pydantic.ValidationError, KeyError) as error:
-            raise BaseConfigurationError(
-                brief="Invalid instance config data.",
-                details=f"Instance configuration data: {data!r}",
-            ) from error
+        :return: The newly created dictionary.
+        """
+        return self.dict(by_alias=True, exclude_unset=True)
+
+    @classmethod
+    def load(
+        cls,
+        executor: Executor,
+        config_path: pathlib.Path = pathlib.Path("/etc/craft-instance.conf"),
+    ) -> Optional["InstanceConfiguration"]:
+        """Load an instance config file from an environment.
+
+        :param executor: Executor for instance.
+        :param config_path: Path to configuration file. Default is `/etc/craft-instance.conf`.
 
-    def save(self, *, executor: Executor, config_path: pathlib.Path) -> None:
-        """Save instance configuration in executed environment.
+        :return: The InstanceConfiguration object or None if the config does not exist or is empty.
+
+        :raise BaseConfigurationError: If the file cannot be loaded from the environment.
+        """
+        with temp_paths.home_temporary_directory() as temp_directory:
+            temp_config_file = temp_directory / "craft-instance.conf"
+            try:
+                executor.pull_file(source=config_path, destination=temp_config_file)
+            except errors.ProviderError as error:
+                raise BaseConfigurationError(
+                    brief=f"Failed to read instance config in environment at {config_path}",
+                ) from error
+            except FileNotFoundError:
+                return None
+            with open(temp_config_file, encoding="utf8") as file:
+                data = yaml.safe_load(file)
+                if data is None:
+                    return None
+
+                return cls.unmarshal(data)
```

Comment:
We can unmarshal to a variable and raise `BaseConfigurationError` here instead of raising it directly from `unmarshal`.